sv.c

```diff
@@ -4205,6 +4205,28 @@ S_sv_buf_to_rw(pTHX_ SV *sv)
       && len                    /* and really is a string */               \
     )
 
+/* The test in this macro was also extracted from Perl_sv_setsv_flags so
+ * that it could be used elsewhere. */
+#ifdef PERL_COPY_ON_WRITE
+#define S_SvPV_shared_hkey_or_CoWable(ssv, dsv, sflags, cur, len)          \
+        (sflags & SVf_IsCOW                                                \
+            ? (!len ||                                                     \
+              ( (CHECK_COWBUF_THRESHOLD(cur,len) || SvLEN(dsv) < cur+1)    \
```

Comment:
IDK the answer id have to compile this brach and do heavy single step to learn an tech correct answer.

Amateur question, shouldn't the bounds check be `SvLEN(dsv) < cur+2)` not `SvLEN(dsv) < cur+1)`?
`dsv` can't cow propogate further unless minimum 2 invisible bytes. 1 invis byte is `'\0'`.

Reply:
Changing the logic of when to COW is not something I want in this PR.

I think you're probably correct, in that if the `CHECK_COWBUF_THRESHOLD(cur,len)` test fails and `SvLEN(dsv) == cur + 1`, a string copy will then occur but the result will not be COWable.

Similarly, `SvGROW(dsv, cur + 1);` in the copying code should probably be `SvGROW(dsv, cur + 2);`

I'll make a note to raise a separate PR for this once this PR has (hopefully) been merged.